rollup_sync_service: invalid memory address or nil pointer dereference

#### System information

Geth version: `scrolltech/l2geth:scroll-v5.8.23`

#### Steps to reproduce the behaviour

upgrade sepolia node to v5.8.23

#### Backtrace

````
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xd4c11e]
goroutine 4267 [running]:
github.com/scroll-tech/go-ethereum/rollup/rollup_sync_service.(*RollupSyncService).updateRollupEvents(0xc003943360, {0xc01a409400?, 0x5, 0x20101004bccab?})
        github.com/scroll-tech/go-ethereum/rollup/rollup_sync_service/rollup_sync_service.go:290 +0x89e
github.com/scroll-tech/go-ethereum/rollup/rollup_sync_service.(*RollupSyncService).fetchRollupEvents(0xc003943360)
        github.com/scroll-tech/go-ethereum/rollup/rollup_sync_service/rollup_sync_service.go:207 +0x20d
github.com/scroll-tech/go-ethereum/rollup/rollup_sync_service.(*RollupSyncService).Start.func1()
        github.com/scroll-tech/go-ethereum/rollup/rollup_sync_service/rollup_sync_service.go:147 +0x1f4
created by github.com/scroll-tech/go-ethereum/rollup/rollup_sync_service.(*RollupSyncService).Start in goroutine 1
        github.com/scroll-tech/go-ethereum/rollup/rollup_sync_service/rollup_sync_service.go:135 +0xcf
````

I started seeing this today with sepolia on 5.8.33 and 5.8.38. It was syncing at tip just fine with no issues on 5.8.33 for the past weeks until today when it started constantly timing out. Looking at the logs it's segfaulting and restarting, and then moving forward again

![Image](https://github.com/user-attachments/assets/e661aeb3-78e6-4bd9-a48f-dda27b599a52)

I tried then upgrading to 5.8.38 with the same results

```
NFO [04-28|23:45:43.041] Imported new chain segment               blocks=1 txs=1 mgas=0.038 elapsed=1.778ms     mgasps=21.619 number=9,342,267 hash=3c2bd3..8ed701 dirty=0.00B
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xd53e7e]

goroutine 409 [running]:
github.com/scroll-tech/go-ethereum/rollup/rollup_sync_service.(*RollupSyncService).updateRollupEvents(0xc00ac62140, {0xc032339500?, 0x5, 0x20101004bccab?})
        github.com/scroll-tech/go-ethereum/rollup/rollup_sync_service/rollup_sync_service.go:299 +0x89e
github.com/scroll-tech/go-ethereum/rollup/rollup_sync_service.(*RollupSyncService).fetchRollupEvents(0xc00ac62140)
        github.com/scroll-tech/go-ethereum/rollup/rollup_sync_service/rollup_sync_service.go:210 +0x20d
github.com/scroll-tech/go-ethereum/rollup/rollup_sync_service.(*RollupSyncService).Start.func1()
        github.com/scroll-tech/go-ethereum/rollup/rollup_sync_service/rollup_sync_service.go:150 +0x1f4
created by github.com/scroll-tech/go-ethereum/rollup/rollup_sync_service.(*RollupSyncService).Start in goroutine 1
        github.com/scroll-tech/go-ethereum/rollup/rollup_sync_service/rollup_sync_service.go:138 +0xcf
INFO [04-28|23:45:43.831] Starting l2geth on Scroll Sepolia testnet... 
INFO [04-28|23:45:43.831] Maximum peer count                       ETH=50 LES=0 total=50
INFO [04-28|23:45:43.831] Smartcard socket not found, disabling    err="stat /run/pcscd/pcscd.comm:
```

It recovers quickly

--- a/params/version.go
+++ b/params/version.go
@@ -24,7 +24,7 @@ import (
 const (
 	VersionMajor = 5         // Major version component of the current release
 	VersionMinor = 8         // Minor version component of the current release
-	VersionPatch = 48        // Patch version component of the current release
+	VersionPatch = 49        // Patch version component of the current release
 	VersionMeta  = "mainnet" // Version metadata to append to the version string
 )
 

--- a/rollup/rollup_sync_service/rollup_sync_service.go
+++ b/rollup/rollup_sync_service/rollup_sync_service.go
@@ -42,11 +42,15 @@ const (
 
 	// defaultLogInterval is the frequency at which we print the latest processed block.
 	defaultLogInterval = 5 * time.Minute
+
+	// rewindL1Height is the number of blocks to rewind the L1 sync height when a missing batch event is detected.
+	rewindL1Height = 10
 )
 
 var (
 	finalizedBlockGauge      = metrics.NewRegisteredGauge("chain/head/finalized", nil)
 	ErrShouldResetSyncHeight = errors.New("ErrShouldResetSyncHeight")
+	ErrMissingBatchEvent     = errors.New("ErrMissingBatchEvent")
 )
 
 // RollupSyncService collects ScrollChain batch commit/revert/finalize events and stores metadata into db.
@@ -223,6 +227,16 @@ func (s *RollupSyncService) fetchRollupEvents() error {
 
 				return nil
 			}
+			if errors.Is(err, ErrMissingBatchEvent) {
+				// If there's a missing batch event, rewind the L1 sync height by some blocks to re-fetch from L1 RPC and
+				// replay creating corresponding CommittedBatchMeta in local DB.
+				// This happens recursively until the missing event has been recovered as we will call fetchRollupEvents again
+				// with the `L1Height = prevL1Height - rewindL1Height`.
+				s.callDataBlobSource.SetL1Height(prevL1Height - rewindL1Height)
+
+				return fmt.Errorf("missing batch event, rewinding L1 sync height by %d blocks: %w", rewindL1Height, err)
+			}
+
 			// Reset the L1 height to the previous value to retry fetching the same data.
 			s.callDataBlobSource.SetL1Height(prevL1Height)
 			return fmt.Errorf("failed to parse and update rollup event logs: %w", err)
@@ -297,12 +311,18 @@ func (s *RollupSyncService) updateRollupEvents(daEntries da.Entries) error {
 				var err error
 				if index > 0 {
 					if parentCommittedBatchMeta, err = rawdb.ReadCommittedBatchMeta(s.db, index-1); err != nil {
-						return fmt.Errorf("failed to read parent committed batch meta, batch index: %v, err: %w", index-1, err)
+						return fmt.Errorf("failed to read parent committed batch meta, batch index: %v, err: %w", index-1, errors.Join(ErrMissingBatchEvent, err))
+					}
+					if parentCommittedBatchMeta == nil {
+						return fmt.Errorf("parent committed batch meta = nil, batch index: %v, err: %w", index-1, ErrMissingBatchEvent)
 					}
 				}
 				committedBatchMeta, err := rawdb.ReadCommittedBatchMeta(s.db, index)
 				if err != nil {
-					return fmt.Errorf("failed to read committed batch meta, batch index: %v, err: %w", index, err)
+					return fmt.Errorf("failed to read committed batch meta, batch index: %v, err: %w", index, errors.Join(ErrMissingBatchEvent, err))
+				}
+				if committedBatchMeta == nil {
+					return fmt.Errorf("committed batch meta = nil, batch index: %v, err: %w\"", index, ErrMissingBatchEvent)
 				}
 
 				chunks, err := s.getLocalChunksForBatch(committedBatchMeta.ChunkBlockRanges)
@@ -447,7 +467,10 @@ func (s *RollupSyncService) getCommittedBatchMeta(commitedBatch da.EntryWithBloc
 	if commitedBatch.Version() == encoding.CodecV7 {
 		parentCommittedBatchMeta, err := rawdb.ReadCommittedBatchMeta(s.db, commitedBatch.BatchIndex()-1)
 		if err != nil {
-			return nil, fmt.Errorf("failed to read parent committed batch meta, batch index: %v, err: %w", commitedBatch.BatchIndex()-1, err)
+			return nil, fmt.Errorf("failed to read parent committed batch meta, batch index: %v, err: %w", commitedBatch.BatchIndex()-1, errors.Join(ErrMissingBatchEvent, err))
+		}
+		if parentCommittedBatchMeta == nil {
+			return nil, fmt.Errorf("parent committed batch meta = nil, batch index: %v, err: %w\"", commitedBatch.BatchIndex()-1, ErrMissingBatchEvent)
 		}
 
 		// If parent batch has a lower version this means this is the first batch of CodecV7.

--- a/rollup/rollup_sync_service/rollup_sync_service.go
+++ b/rollup/rollup_sync_service/rollup_sync_service.go
@@ -322,7 +322,7 @@ func (s *RollupSyncService) updateRollupEvents(daEntries da.Entries) error {
 					return fmt.Errorf("failed to read committed batch meta, batch index: %v, err: %w", index, errors.Join(ErrMissingBatchEvent, err))
 				}
 				if committedBatchMeta == nil {
-					return fmt.Errorf("committed batch meta = nil, batch index: %v, err: %w\"", index, ErrMissingBatchEvent)
+					return fmt.Errorf("committed batch meta = nil, batch index: %v, err: %w", index, ErrMissingBatchEvent)
 				}
 
 				chunks, err := s.getLocalChunksForBatch(committedBatchMeta.ChunkBlockRanges)
@@ -470,7 +470,7 @@ func (s *RollupSyncService) getCommittedBatchMeta(commitedBatch da.EntryWithBloc
 			return nil, fmt.Errorf("failed to read parent committed batch meta, batch index: %v, err: %w", commitedBatch.BatchIndex()-1, errors.Join(ErrMissingBatchEvent, err))
 		}
 		if parentCommittedBatchMeta == nil {
-			return nil, fmt.Errorf("parent committed batch meta = nil, batch index: %v, err: %w\"", commitedBatch.BatchIndex()-1, ErrMissingBatchEvent)
+			return nil, fmt.Errorf("parent committed batch meta = nil, batch index: %v, err: %w", commitedBatch.BatchIndex()-1, ErrMissingBatchEvent)
 		}
 
 		// If parent batch has a lower version this means this is the first batch of CodecV7.

--- a/rollup/rollup_sync_service/rollup_sync_service.go
+++ b/rollup/rollup_sync_service/rollup_sync_service.go
@@ -44,7 +44,7 @@ const (
 	defaultLogInterval = 5 * time.Minute
 
 	// rewindL1Height is the number of blocks to rewind the L1 sync height when a missing batch event is detected.
-	rewindL1Height = 10
+	rewindL1Height = 100
 )
 
 var (
@@ -234,7 +234,7 @@ func (s *RollupSyncService) fetchRollupEvents() error {
 				// with the `L1Height = prevL1Height - rewindL1Height`.
 				s.callDataBlobSource.SetL1Height(prevL1Height - rewindL1Height)
 
-				return fmt.Errorf("missing batch event, rewinding L1 sync height by %d blocks: %w", rewindL1Height, err)
+				return fmt.Errorf("missing batch event, rewinding L1 sync height by %d blocks to %d: %w", rewindL1Height, prevL1Height-rewindL1Height, err)
 			}
 
 			// Reset the L1 height to the previous value to retry fetching the same data.

--- a/rollup/rollup_sync_service/rollup_sync_service.go
+++ b/rollup/rollup_sync_service/rollup_sync_service.go
@@ -228,13 +228,19 @@ func (s *RollupSyncService) fetchRollupEvents() error {
 				return nil
 			}
 			if errors.Is(err, ErrMissingBatchEvent) {
+				// make sure no underflow
+				var rewindTo uint64
+				if prevL1Height > rewindL1Height {
+					rewindTo = prevL1Height - rewindL1Height
+				}
+
 				// If there's a missing batch event, rewind the L1 sync height by some blocks to re-fetch from L1 RPC and
 				// replay creating corresponding CommittedBatchMeta in local DB.
 				// This happens recursively until the missing event has been recovered as we will call fetchRollupEvents again
 				// with the `L1Height = prevL1Height - rewindL1Height`.
-				s.callDataBlobSource.SetL1Height(prevL1Height - rewindL1Height)
+				s.callDataBlobSource.SetL1Height(rewindTo)
 
-				return fmt.Errorf("missing batch event, rewinding L1 sync height by %d blocks to %d: %w", rewindL1Height, prevL1Height-rewindL1Height, err)
+				return fmt.Errorf("missing batch event, rewinding L1 sync height by %d blocks to %d: %w", rewindL1Height, rewindTo, err)
 			}
 
 			// Reset the L1 height to the previous value to retry fetching the same data.

